### PR TITLE
feat: handle function and contract resolve with alias

### PR DESCRIPTION
**Pull request by Paulalex85**
The plugin was not able to resolve when an alias is used in the import

### CODE REVIEW

File: src/test/kotlin/me/serce/solidity/lang/core/resolve/SolAliasResolveTest.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      """,
+            name = "a.sol"
+        ),
+        InlineFile(
+            """
+          import {a as A} from "./a.sol";
+              
+          contract b {
+            function test() {
+                A.doit();
```
Comment (mbolotov) [inline src/test/kotlin/me/serce/solidity/lang/core/resolve/SolAliasResolveTest.kt:21]: this example compiles by solc with error: `TypeError: Cannot call function via contract type name.`
Let's change it (and other tests) to make a real use case.

File: src/main/kotlin/me/serce/solidity/lang/resolve/engine.kt [hunk 2]
```diff
@@ -322,13 +322,44 @@ object SolResolver {
...
+              //match import a as A from "path"
+              if (it.userDefinedTypeName != null) {
+                nameToFilter = it.userDefinedTypeName!!.name
+                return@run
+              } else {
+              //match import "path" as A
+                val contractFromPath =
+                  it.importPath?.reference?.resolve()?.childrenOfType<SolContractDefinition>()
+                if (!contractFromPath.isNullOrEmpty()) {
+                  nameToFilter = contractFromPath[0].name
```
Comment (mbolotov) [inline src/main/kotlin/me/serce/solidity/lang/resolve/engine.kt:340]: why do we take the first contract only? there may be multiple contracts in a file

**Review by mbolotov — CHANGES_REQUESTED** [review]

**Comment by mbolotov** [comment]
Now tests are failing. Please check.

File: src/test/kotlin/me/serce/solidity/lang/core/resolve/SolAliasResolveTest.kt (hunk 1 above)
Comment (Paulalex85) [reply src/test/kotlin/me/serce/solidity/lang/core/resolve/SolAliasResolveTest.kt:25]: Yes thank you for this feedback, I found more use cases. I'm gonna fix them

File: src/main/kotlin/me/serce/solidity/lang/resolve/engine.kt [hunk 3]
```diff
@@ -180,6 +180,14 @@ object SolResolver {
     return result + result.map { collectImports(it.file.childrenOfType<SolImportDirective>(), visited) }.flatten()
   }
 
+  //collect all SolContractDefinition recursively from imports
+  fun collectContracts(file: PsiFile): Collection<SolContractDefinition> {
```
Comment (mbolotov) [inline src/main/kotlin/me/serce/solidity/lang/resolve/engine.kt:184]: please merge the master into your branch.
there was `collectImports` method added within https://github.com/intellij-solidity/intellij-solidity/pull/450
it looks like that method do the same thing, but it also contains a recursive call guard

Comment (Paulalex85) [reply src/main/kotlin/me/serce/solidity/lang/resolve/engine.kt]: Indeed it's much better, thanks !
I already merged main into my branch

**Review by mbolotov — APPROVED** [review]

**Review by SerCeMan — APPROVED** [review]
Thanks, @Paulalex85!

**Pull request merged**